src/Backend/API.ts

```diff
@@ -24,32 +31,299 @@
-  /* TODO Implement */
+class Command extends Array<string> {
+  readonly cmd: string;
+  options?: string[];
+
+  constructor(cmd: string, options?: string[]) {
+    super();
+    this.cmd = cmd;
+    this.options = options;
+  }
+
+  strs(): string[] {
+    let ret: string[] = [this.cmd];
+    if (this.options !== undefined) {
+      ret.concat(this.options);
+    }
+    return ret;
+  }
+
+  str(): string {
+    return this.strs().join(' ');
+  }
 };
 
-/**
- * Interface of backend extension
- */
-interface Backend {
-  /**
-   * Return name of backend
-   */
-  name(): string;
+export class JobToolchain extends JobBase {
+  cmd: Command;
+  constructor(type: Job.Type, cmd: Command) {
+    super();
+    this.jobType = type;
+    this.cmd = cmd;
+
+    assert(this.jobType >= Job.Type.tInstall && this.jobType <= Job.Type.tList);
+    assert(this.cmd.strs.length >= 3);
+  }
+
+  public get valid() {
+    // TODO validate arguments;
+    return true;
+  }
+
+  public get driver() {
+    return this.cmd.strs()[0];
+  }
 
-  //
-  // Compiler toolchain API
-  //
+  public get tool() {
+    return this.cmd.strs()[1];
+  }
+
+  public get toolArgs() {
+    let args = new ToolArgs();
+    args.concat(this.cmd.strs().slice(2));
+    return args;
+  }
+};
+
+export class JobInstall extends JobToolchain {
+  constructor(cmd: Command) {
+    super(Job.Type.tInstall, cmd);
+  }
+};
 
-  /**
-   * Search already-installed compiler toolchain in local desktop and remote server
-   */
-  getInstalledCompilerToolchain(): CompilerToolchains;
+export class JobUninstall extends JobToolchain {
+  constructor(cmd: Command) {
+    super(Job.Type.tUninstall, cmd);
+  }
+};
+
+export class JobList extends JobToolchain {
+  constructor(cmd: Command) {
+    super(Job.Type.tList, cmd);
+  }
+};
+
+export class JobConfig extends JobBase {
+  cfgFile: string;
+
+  constructor(cfgFile: string) {
+    super();
+    this.jobType = Job.Type.tConfig;
+    this.cfgFile = cfgFile;
+  }
+
+  public get valid() {
+    // TODO validate arguments;
+    return true;
+  }
+
+  public get tool() {
+    return '';
+  }
+
+  public get toolArgs() {
+    let args = new ToolArgs();
+    args.add('--config', this.cfgFile);
+    return args;
+  }
+};
+
+// TODO: Version can be other variations such as `1.0.0~220415`.
+class Version {
+  major: number;
+  minor: number;
+  patch: number;
+
+  constructor(major: number = 0, minor: number = 0, patch: number = 0) {
+    this.major = major;
+    this.minor = minor;
+    this.patch = patch;
+  }
+
+  str(): string {
+    let ret: string = `${this.major}.${this.minor}.${this.patch}`;
+    return ret;
+  }
+
+  // TODO
+  // operator <, =, >
+};
+
+export interface ToolchainInfo {
+  name: string;
+  description: string;
+  version?: Version;         // undefined means latest
+  minimumVersion?: Version;  // undefined means all
+}
 
-  // TODO Define more functions
+// Toolchain: Debian(...) OR Docker(...)
+export interface Toolchain {
```

Comment:
Can I ask a question?
toolchain could be 
- type 1. the one in artifactory (zipped installable file)
- type 2. the one already-installed in my Ubuntu PC

`Compiler.toolchains()` returns list of toolchain.
I wonder whether returned toolchains are the list of type 1 or type 2.
I guess I think this should return type 2.

Then the method here seems not natural since calling `install()` of alrady-installed toolchain is not correct.
Should these methods (`install(), uninstall(), list()`) be `static` ?

Reply:
I think this question is resolved.